### PR TITLE
[opt] Simplify replace_statements and improve demote_dense_struct_fors

### PR DESCRIPTION
**Pull request by xumingkuan**
Related issue = #1989 -- This PR provides a filtered map operation on the IR. We can replace some `gather_statements` usages with `transform_statements` or something similar.

A continuation of #2306.

Also reduces the time complexity of demote_dense_struct_fors from O(snodes.size() * (physical_indices.size() + N)) to O(snodes.size() * physical_indices.size() + N) where N is the total number of statements in the IR.

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

File: taichi/ir/transforms.h [hunk 1]
```diff
@@ -75,25 +75,29 @@ bool constant_fold(IRNode *root,
...
 void offload(IRNode *root, const CompileConfig &config);
+bool transform_statements(
+    IRNode *root,
+    std::function<bool(Stmt *)> filter,
+    std::function<void(Stmt *, DelayedIRModifier *)> transformer);
 /**
  * @param root The IR root to be traversed.
  * @param filter A function which tells if a statement need to be replaced.
  * @param generator If a statement s need to be replaced, generate a new
  * statement s1 with the argument s, insert s1 to s's place, and replace all
```
Comment (k-ye) [inline taichi/ir/transforms.h:86]: nit: by saying "insert s1 to s's place", do we mean "insert |s1| to where |s| is defined, removes |s|'s definition and replaces all usages of ..."?

File: taichi/ir/transforms.h [hunk 2]
```diff
@@ -75,25 +75,29 @@ bool constant_fold(IRNode *root,
...
 /**
- * @param generator If a statement s need to be replaced, find the existing
+ * @param finder If a statement s need to be replaced, find the existing
  * statement s1 with the argument s, and replace all usages of s with s1.
- * @return Whether the IR is modified.
  */
-bool replace_statements_with(IRNode *root,
-                             std::function<bool(Stmt *)> filter,
-                             std::function<Stmt *(Stmt *)> generator);
+bool replace_statements(IRNode *root,
```
Comment (k-ye) [inline taichi/ir/transforms.h:98]: Following the above comment, how is |s|'s defining stmt handled in this case?

**Review by k-ye — APPROVED** [review]
LGTM!

File: taichi/ir/transforms.h (hunk 1 above)
Comment (xumingkuan) [reply taichi/ir/transforms.h:86]: Right, I'll rephrase this.

File: taichi/ir/transforms.h (hunk 2 above)
Comment (xumingkuan) [reply taichi/ir/transforms.h:98]: |s|'s defining stmt is erased here.

**Pull request merged**